### PR TITLE
Add `pmap` compatibility for JAX `data_iterator`

### PR DESCRIPTION
**Pull request by awolant**
## Category:

**New feature** 

## Description:

This PR adds new syntax to `data_iterator`: new argument called `devices`. It can be used to configure the iterator to return output that is compatible with `pmap`ed functions in JAX. 

## Additional information:

JAX has to ways of spreading the workload to multiple GPUs: automatic parallelization and `pmap` function. `data_iterator` supported automatic parallelization with `sharding` argument. For `pmap` support user had to manually create multiple pipelines and pass them to `DALIGenericIterator`. In this PR I added a way to use `data_iterator` with `pmap`.

With this introduction, I reworked tutorial for multigpu for DALI and JAX to reflect this change. While at it, I edited this tutorial to better reflect desired priority: automatic parallelization first, `pmap` second.

### Affected modules and functionalities:

`data_iterator` and all its related code. I added some minor changes, bug fixes etc. to `data_iterator` and its tests. 

### Key points relevant for the review:

`devices` arg: is it ok? tutorial: is it ok?

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [x] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3710

### CODE REVIEW

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/jax/iterator.py [hunk 1]
```diff
@@ -252,54 +252,101 @@ def data_iterator_impl(
         last_batch_padded=False,
         last_batch_policy=LastBatchPolicy.FILL,
         prepare_first_batch=True,
-        sharding=None):
+        sharding=None,
+        devices=None):
```
Comment (awolant) [inline dali/python/nvidia/dali/plugin/jax/iterator.py:256]: This is the place where actual functionality is added. This function is a common implementation used in different decorators later.

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/jax/clu.py [hunk 2]
```diff
@@ -294,9 +295,15 @@ def peekable_data_iterator(
...
-                build an output jax.Array for each category. If ``None``, the iterator returns
-                values compatible with pmapped JAX functions.
+    sharding : ``jax.sharding.Sharding`` compatible object that, if present, will be used to
+                build an output jax.Array for each category. Iterator will return outputs
+                compatible with automatic parallelization in JAX.
+                This argument is mutually exclusive with `devices` argument. If `devices` is
+                provided, `sharding` should be set to None.
+    devices : list of jax.devices to be used to run the pipeline in parallel. Iterator will
+                return outputs compatible with pmapped JAX functions.
+                This argument is  mutually exclusive with `sharding` argument. If `sharding`
```
Comment (stiepan) [inline dali/python/nvidia/dali/plugin/jax/clu.py:305]:
```suggestion
                This argument is mutually exclusive with `sharding` argument. If `sharding`
```

File: docs/examples/frameworks/jax/jax-multigpu_example.ipynb [hunk 3]
```diff
@@ -7,27 +7,58 @@
    "source": [
```
Comment (stiepan) [inline docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <blockquote>Here we show how to run training from <a href="https://github.com/NVIDIA/DALI/raw/b5fade136427602a13eb071d7c560a31cea2df47/docs/examples/frameworks/jax/jax-basic_example.ipynb" target="_blank">Training neural network with DALI and JAX</a> on multiple GPUs. If you haven't already done so it is best to start with single GPU example to better understand following content.  </blockquote><p>I may be picky or just drowsy, but I read <code>If you haven't already done so</code> as if referring to training NN on multiple GPUs. How about something like:</p><blockquote>Here we show how to run training from <a href="https://github.com/NVIDIA/DALI/raw/b5fade136427602a13eb071d7c560a31cea2df47/docs/examples/frameworks/jax/jax-basic_example.ipynb" target="_blank">Training neural network with DALI and JAX</a> on multiple GPUs. It is best to start with single GPU example, if you haven't already done so.</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/5185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (stiepan) [inline docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <blockquote>that we want to apply to <s>out</s> (our?) computation.</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/5185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (stiepan) [inline docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>on <strong>a</strong> single GPU?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/5185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (stiepan) [inline docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>on <strong>a</strong> single GPU?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/5185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

**Review by stiepan — APPROVED** [review]

File: docs/examples/frameworks/jax/jax-multigpu_example.ipynb (hunk 3 above)
Comment (awolant) [reply docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>Done. Changed the whole thing</p>

Comment (awolant) [reply docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>Done</p>

Comment (awolant) [reply docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>Done</p>

Comment (awolant) [reply docs/examples/frameworks/jax/jax-multigpu_example.ipynb:7]: <p>Done</p>

File: dali/python/nvidia/dali/plugin/jax/clu.py (hunk 2 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/jax/clu.py]: Done

**Review by banasraf — APPROVED** [review]

**Comment by awolant** [comment]
!build

**Pull request merged**